Model persistence breaks with impersonation

### Describe the bug

Yup, you read it right, no models will be persisted if impersonation is enabled

### To Reproduce

1) run the following query in your postgres

<details>

```
CREATE ROLE something;

GRANT SELECT ON ALL TABLES IN SCHEMA PUBLIC TO something;


create table public."table" (
    usage_date date not null,
    model varchar(10) not null,
    device_id varchar(255) not null
);

insert into "table" (usage_date, model, device_id)
values
    ('2025-01-01', '1', 1245),
    ('2025-01-01', '2', 842),
    ('2025-01-01', '3', 391),
    ('2025-02-01', '1', 1310),
    ('2025-02-01', '2', 876),
    ('2025-02-01', '3', 402),
    ('2025-03-01', '1', 1388),
    ('2025-03-01', '2', 921),
    ('2025-03-01', '3', 417),
    ('2025-04-01', '1', 1420),
    ('2025-04-01', '2', 957),
    ('2025-04-01', '3', 433),
    ('2025-05-01', '1', 1495),
    ('2025-05-01', '2', 1003),
    ('2025-05-01', '3', 451),
    ('2025-06-01', '1', 1532),
    ('2025-06-01', '2', 1044),
    ('2025-06-01', '3', 468),
    ('2025-07-01', '1', 1598),
    ('2025-07-01', '2', 1087),
    ('2025-07-01', '3', 489),
    ('2025-08-01', '1', 1640),
    ('2025-08-01', '2', 1125),
    ('2025-08-01', '3', 501),
    ('2025-09-01', '1', 1712),
    ('2025-09-01', '2', 1178),
    ('2025-09-01', '3', 526),
    ('2025-10-01', '1', 1769),
    ('2025-10-01', '2', 1210),
    ('2025-10-01', '3', 548),
    ('2025-11-01', '1', 1815),
    ('2025-11-01', '2', 1259),
    ('2025-11-01', '3', 571),
    ('2025-12-01', '1', 1884),
    ('2025-12-01', '2', 1302),
    ('2025-12-01', '3', 596);
```

</details>

2) create a group in metabase and create a user, add attribute role = something

3) enable model caching (go to performance-> enable model caching, then to database and enable model persistence)

4) create a SQL query in Metabase as an admin

```
SELECT
      DATE_TRUNC('MONTH', usage_date) AS usage_month,
      CASE
        WHEN model = 'plus' THEN '1'
        WHEN model = 'x' THEN '2'
        WHEN model = 'm' THEN '3'
      END AS model_,
      COUNT(DISTINCT device_id) AS n_devices
    FROM
      "table"
    WHERE
      usage_date >= CAST(CURRENT_TIMESTAMP AS DATE) - 600
      AND model <> '0'
    GROUP BY
      usage_month,
      model_
    ORDER BY
      usage_month,
      model_
```

SAVE THIS QUERY AS A MODEL, now go to the settings of the model and see the failure
 

### Expected behavior

Models should be able to be cached

### Logs

<details>
2026-05-15 17:17:21,166 INFO task.persist-refresh :: Error refreshing persisting model with card-id 77
clojure.lang.ExceptionInfo: Error preprocessing query in #'metabase.query-processor.middleware.enterprise/apply-impersonation: Invalid impersonated native query. Must be a single insert, update, or delete statement. {:fn #'metabase.query-processor.middleware.enterprise/apply-impersonation, :query {:lib/type :mbql/query, :lib/metadata (metabase.lib.metadata.invocation-tracker/invocation-tracker-provider (metabase.lib.metadata.cached-provider/cached-metadata-provider (metabase.lib-be.metadata.jvm/->UncachedApplicationDatabaseMetadataProvider 1))), :database 1, :stages [{:lib/type :mbql.stage/native, :native "select\n\tdate_trunc('month', usage_date) as usage_month,\n\tcase\n\t\twhen model = 'plus' then '1'\n\t\twhen model = 'x' then '2'\n\t\twhen model = 'm' then '3'\n\t\tend as model_,\n\tcount(distinct device_id) as n_devices\nfrom\n\t\"table\"\nwhere\n\tusage_date >= now()::date - 600\n\tand model != '0'\ngroup by\n\tusage_month,\n\tmodel_\norder by\n\tusage_month,\n\tmodel_\n;\n"}], :middleware {:disable-max-results? true}}, :type :qp}
        at metabase.query_processor.preprocess$preprocess$fn__83216$fn__83217.invoke(preprocess.clj:149)
        at clojure.lang.PersistentVector.reduce(PersistentVector.java:418)
        at clojure.core$transduce.invokeStatic(core.clj:7025)
        at clojure.core$transduce.invoke(core.clj:7012)
        at metabase.query_processor.preprocess$preprocess$fn__83216.invoke(preprocess.clj:123)
        at metabase.query_processor.setup$do_with_qp_setup.invokeStatic(setup.clj:232)
        at metabase.query_processor.setup$do_with_qp_setup.invoke(setup.clj:223)
        at metabase.query_processor.preprocess$preprocess.invokeStatic(preprocess.clj:121)
        at metabase.query_processor.preprocess$preprocess.invoke(preprocess.clj:116)
        at metabase.query_processor.compile$compile$fn__83587.invoke(compile.clj:66)
        at metabase.query_processor.setup$do_with_canceled_chan$fn__82816.invoke(setup.clj:196)
        at metabase.query_processor.setup$do_with_database_local_settings$fn__82811.invoke(setup.clj:188)
        at metabase.query_processor.setup$do_with_driver$fn__82798$fn__82801.invoke(setup.clj:173)
        at metabase.driver$do_with_driver.invokeStatic(driver.clj:54)
        at metabase.driver$do_with_driver.invoke(driver.clj:49)
        at metabase.query_processor.setup$do_with_driver$fn__82798.invoke(setup.clj:172)
        at metabase.query_processor.setup$do_with_metadata_provider$fn__82790$fn__82791.invoke(setup.clj:141)
        at metabase.query_processor.store$do_with_metadata_provider.invokeStatic(store.clj:200)
        at metabase.query_processor.store$do_with_metadata_provider.invoke(store.clj:186)
        at metabase.query_processor.store$do_with_metadata_provider.invokeStatic(store.clj:193)
        at metabase.query_processor.store$do_with_metadata_provider.invoke(store.clj:186)
        at metabase.query_processor.setup$do_with_metadata_provider$fn__82790.invoke(setup.clj:140)
        at metabase.query_processor.setup$do_with_resolved_database$fn__82782.invoke(setup.clj:125)
        at metabase.query_processor.setup$do_with_qp_setup.invokeStatic(setup.clj:240)
        at metabase.query_processor.setup$do_with_qp_setup.invoke(setup.clj:223)
        at metabase.query_processor.compile$compile.invokeStatic(compile.clj:65)
        at metabase.query_processor.compile$compile.invoke(compile.clj:62)
        at metabase.driver.postgres.ddl$fn__111534.invokeStatic(ddl.clj:39)
        at metabase.driver.postgres.ddl$fn__111534.invoke(ddl.clj:37)
        at clojure.lang.MultiFn.invoke(MultiFn.java:244)
        at metabase.model_persistence.task.persist_refresh$reify__155141.refresh_BANG_(persist_refresh.clj:52)
        at metabase.model_persistence.task.persist_refresh$refresh_with_stats_BANG_$fn__155157.invoke(persist_refresh.clj:76)
        at metabase.model_persistence.task.persist_refresh$refresh_with_stats_BANG_.invokeStatic(persist_refresh.clj:75)
        at metabase.model_persistence.task.persist_refresh$refresh_with_stats_BANG_.invoke(persist_refresh.clj:56)
        at clojure.lang.AFn.applyToHelper(AFn.java:165)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:673)
        at clojure.core$partial$fn__5935.doInvoke(core.clj:2660)
        at clojure.lang.RestFn.invoke(RestFn.java:400)
        at metabase.model_persistence.task.persist_refresh$save_task_history_BANG_$fn__155203.invoke(persist_refresh.clj:127)
        at metabase.task_history.models.task_history$do_with_task_history.invokeStatic(task_history.clj:238)
        at metabase.task_history.models.task_history$do_with_task_history.invoke(task_history.clj:221)
        at metabase.model_persistence.task.persist_refresh$save_task_history_BANG_.invokeStatic(persist_refresh.clj:119)
        at metabase.model_persistence.task.persist_refresh$save_task_history_BANG_.invoke(persist_refresh.clj:115)
        at metabase.model_persistence.task.persist_refresh$refresh_individual_BANG_.invokeStatic(persist_refresh.clj:231)
        at metabase.model_persistence.task.persist_refresh$refresh_individual_BANG_.invoke(persist_refresh.clj:223)
        at metabase.model_persistence.task.persist_refresh$refresh_job_fn_BANG_.invokeStatic(persist_refresh.clj:247)
        at metabase.model_persistence.task.persist_refresh$refresh_job_fn_BANG_.invoke(persist_refresh.clj:241)
        at metabase.model_persistence.task.persist_refresh.PersistenceRefresh$fn__155302$fn__155303.invoke(persist_refresh.clj:257)
        at metabase.util.log$with_exception_context_fn.invokeStatic(log.clj:350)
        at metabase.util.log$with_exception_context_fn.invoke(log.clj:342)
        at metabase.model_persistence.task.persist_refresh.PersistenceRefresh$fn__155302.invoke(persist_refresh.clj:254)
        at metabase.util.log$with_thread_context_fn.invokeStatic(log.clj:88)
        at metabase.util.log$with_thread_context_fn.invoke(log.clj:74)
        at metabase.model_persistence.task.persist_refresh.PersistenceRefresh.execute(persist_refresh.clj:254)
        at org.quartz.core.JobRunShell.run(JobRunShell.java:202)
        at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:573)
Caused by: clojure.lang.ExceptionInfo: Invalid impersonated native query. Must be a single insert, update, or delete statement. {:type :invalid-query, :sql "select\n\tdate_trunc('month', usage_date) as usage_month,\n\tcase\n\t\twhen model = 'plus' then '1'\n\t\twhen model = 'x' then '2'\n\t\twhen model = 'm' then '3'\n\t\tend as model_,\n\tcount(distinct device_id) as n_devices\nfrom\n\t\"table\"\nwhere\n\tusage_date >= now()::date - 600\n\tand model != '0'\ngroup by\n\tusage_month,\n\tmodel_\norder by\n\tusage_month,\n\tmodel_\n;\n"}
        at metabase.driver.sql$validate_impersonated_query_STAR_$fn__99973$fn__99974.invoke(sql.clj:267)
        at metabase.util.performance$apply_and_small_conj_BANG_.invokeStatic(performance.cljc:173)
        at metabase.util.performance$apply_and_small_conj_BANG_.invoke(performance.cljc:171)
        at metabase.util.performance$reduce.invokeStatic(performance.cljc:99)
        at metabase.util.performance$reduce.invoke(performance.cljc:89)
        at metabase.util.performance$mapv.invokeStatic(performance.cljc:208)
        at metabase.util.performance$mapv.invoke(performance.cljc:200)
        at metabase.driver.sql$validate_impersonated_query_STAR_$fn__99973.invoke(sql.clj:253)
        at clojure.core$update.invokeStatic(core.clj:6259)
        at clojure.core$update.invoke(core.clj:6251)
        at metabase.driver.sql$validate_impersonated_query_STAR_.invokeStatic(sql.clj:251)
        at metabase.driver.sql$validate_impersonated_query_STAR_.invoke(sql.clj:245)
        at metabase.driver.postgres$fn__112418.invokeStatic(postgres.clj:1438)
        at metabase.driver.postgres$fn__112418.invoke(postgres.clj:1436)
        at clojure.lang.MultiFn.invoke(MultiFn.java:234)
        at metabase_enterprise.impersonation.middleware$fn__174871$apply_impersonation__174872.invoke(middleware.clj:26)
        at clojure.lang.AFn.applyToHelper(AFn.java:154)
        at clojure.lang.AFn.applyTo(AFn.java:144)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$apply.invoke(core.clj:662)
        at metabase.premium_features.defenterprise$dynamic_ee_oss_fn$fn__54839.doInvoke(defenterprise.clj:54)
        at clojure.lang.RestFn.invoke(RestFn.java:411)
        at clojure.lang.Var.invoke(Var.java:386)
        at metabase.query_processor.preprocess$preprocess$fn__83216$fn__83217.invoke(preprocess.clj:134)
        ... 56 more
2026-05-15 17:17:21,327 INFO task.persist-refresh :: Finished updated model-id 77 from persisted-info 1.
</details>

### Information about your Metabase installation

```JSON
60.6
```

### Severity

P1

### Additional context

I'm crying writing this bug report

--- a/enterprise/backend/test/metabase_enterprise/impersonation/driver_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/impersonation/driver_test.clj
@@ -767,6 +767,34 @@
                                        "DROP ROLE IF EXISTS \"impersonation_role\";"]]
                       (jdbc/execute! spec [statement]))))))))))))
 
+(deftest native-model-persistence-works-when-impersonation-enabled-test
+  ;; Test explicitly with postgres since it supports persistence, impersonation, and impersonated native-query
+  ;; validation.
+  (mt/test-driver :postgres
+    (mt/with-premium-features #{:advanced-permissions}
+      (mt/dataset test-data
+        (mt/with-persistence-enabled! [persist-models!]
+          (impersonation.util-test/with-impersonations! {:impersonations [{:db-id (mt/id) :attribute "impersonation_attr"}]
+                                                         :attributes     {"impersonation_attr" "impersonation_role"}}
+            (request/as-admin
+              (mt/with-temp [:model/Card model {:type            :model
+                                                :database_id     (mt/id)
+                                                :query_type      :native
+                                                :dataset_query   (mt/native-query {:query "SELECT 1 AS id"})
+                                                :result_metadata [{:name      "id"
+                                                                   :base_type :type/Integer}]}]
+                (try
+                  (persist-models!)
+                  (catch Exception e
+                    (tap> e)
+                    (throw e)))
+                (is (=? {:state  "persisted"
+                         :active true
+                         :error  nil}
+                        (t2/select-one [:model/PersistedInfo :state :active :error]
+                                       :database_id (mt/id)
+                                       :card_id (:id model))))))))))))
+
 (deftest resilient-connection-options-test
   (testing "resilient connections have the correct role set"
     (mt/test-drivers (mt/normal-driver-select {:+parent :sql-jdbc

--- a/enterprise/backend/test/metabase_enterprise/impersonation/driver_test.clj
+++ b/enterprise/backend/test/metabase_enterprise/impersonation/driver_test.clj
@@ -783,11 +783,7 @@
                                                 :dataset_query   (mt/native-query {:query "SELECT 1 AS id"})
                                                 :result_metadata [{:name      "id"
                                                                    :base_type :type/Integer}]}]
-                (try
-                  (persist-models!)
-                  (catch Exception e
-                    (tap> e)
-                    (throw e)))
+                (persist-models!)
                 (is (=? {:state  "persisted"
                          :active true
                          :error  nil}

--- a/src/metabase/driver/connection.clj
+++ b/src/metabase/driver/connection.clj
@@ -69,7 +69,7 @@
    - `:default` — primary `:details`
    - `:write-data` — `:write-data-details` merged over `:details` (if configured)
 
-   Bind via [[with-write-connection]], not directly."
+   Bind via [[with-write-connection]] or [[with-default-connection]], not directly."
   :default)
 
 (defmacro with-write-connection
@@ -79,6 +79,14 @@
    into account (if configured) instead of only primary `:details`."
   [& body]
   `(binding [*connection-type* :write-data]
+     ~@body))
+
+(defmacro with-default-connection
+  "Establishes a default-connection context for body.
+
+   Use this to compile or execute a read query from inside a broader write-connection context."
+  [& body]
+  `(binding [*connection-type* :default]
      ~@body))
 
 (def ^:dynamic ^:private *suppress-resolution-telemetry*

--- a/src/metabase/driver/mysql/ddl.clj
+++ b/src/metabase/driver/mysql/ddl.clj
@@ -6,6 +6,7 @@
    [honey.sql :as sql]
    [java-time.api :as t]
    [metabase.driver-api.core :as driver-api]
+   [metabase.driver.connection :as driver.conn]
    [metabase.driver.ddl.interface :as ddl.i]
    [metabase.driver.sql-jdbc.connection :as sql-jdbc.conn]
    [metabase.driver.sql-jdbc.execute :as sql-jdbc.execute]
@@ -67,7 +68,8 @@
 
 (defmethod ddl.i/refresh! :mysql
   [driver database definition dataset-query]
-  (let [{:keys [query params]} (driver-api/compile dataset-query)
+  (let [{:keys [query params]} (driver.conn/with-default-connection
+                                 (driver-api/compile dataset-query))
         db-spec (sql-jdbc.conn/db->pooled-connection-spec database)]
     (sql-jdbc.execute/do-with-connection-with-options
      driver

--- a/src/metabase/driver/postgres/ddl.clj
+++ b/src/metabase/driver/postgres/ddl.clj
@@ -4,6 +4,7 @@
    [honey.sql :as sql]
    [java-time.api :as t]
    [metabase.driver-api.core :as driver-api]
+   [metabase.driver.connection :as driver.conn]
    [metabase.driver.ddl.interface :as ddl.i]
    [metabase.driver.sql-jdbc.execute :as sql-jdbc.execute]
    [metabase.driver.sql.ddl :as sql.ddl]
@@ -36,7 +37,8 @@
 
 (defmethod ddl.i/refresh! :postgres
   [driver database definition dataset-query]
-  (let [{:keys [query params]} (driver-api/compile dataset-query)]
+  (let [{:keys [query params]} (driver.conn/with-default-connection
+                                 (driver-api/compile dataset-query))]
     (sql-jdbc.execute/do-with-connection-with-options
      driver
      database

